Replace LRR module with generic RSS module

HubbeKing opened the pull request:
Closes #36, assuming it works? Should. The admin-checking is maybe a bit wonk?

HubbeKing committed:
Add generic RSS module

HubbeKing committed:
Delete LRR-specific RSS module

we'll have to manually .rss add the LRR feeds if we care

StarlitGhost commented on desertbot/modules/commands/RSS.py:
```diff
@@ -0,0 +1,184 @@
+    def _listFeeds(self, message: IRCMessage):
+        return IRCResponse(ResponseType.Say,
+                           "Currently followed feeds are: {}".format(", ".join(self.feeds.keys())),
+                           message.replyTo)
```
you're missing the rss = RSS() at the bottom :P

HubbeKing replied:
oh I am a smart

HubbeKing committed:
Actually store feed data after update

HubbeKing committed:
Helps if you actually instantiate the class

StarlitGhost commented on desertbot/modules/commands/RSS.py:
```diff
@@ -0,0 +1,188 @@
+                "lastCheck": datetime.datetime.utcnow() - datetime.timedelta(minutes=10),
+                "suppress": False
+            }
+            self.feeds.append({name: feed_object})
```
dicts don't have an append method, I guess you just want self.feeds[name] = feed_object

HubbeKing committed:
Dicts don't have an .append

Don't code when tired, kids.

HubbeKing commented:
Travis build failure is unrelated, see #146

HubbeKing committed:
Have checkFeeds call execute() if it's an RSS command

HubbeKing committed:
Move all handling code to checkFeeds

execute still gets called after checkFeeds, so there was double output

HubbeKing committed:
It's last, not latest

StarlitGhost commented on desertbot/modules/commands/RSS.py:
```diff
@@ -0,0 +1,189 @@
+                feed = message.parameters.strip()
+                latest = self._getLatest(feed)
+                if latest is not None:
+                    response = 'Latest {}: {} | {}'.format(feed["name"], feed["title"], feed["link"])
```
`latest`, not `feed`

HubbeKing committed:
latest, not feed

HubbeKing committed:
THERE MUST BE ORDER

HubbeKing committed:
datetime objects aren't JSON serializable

ISO8601 strings should compare just as easily as datetime objects, though

Heufneutje commented on desertbot/modules/commands/RSS.py:
```diff
@@ -0,0 +1,189 @@
+                self.feeds[feedName]["lastUpdate"] = newestDate
+                title = item.find("title").text
+                link = item.find("link").text
+                link = self.bot.moduleHandler.runActionUntilValue("shorten-url", link)
```
This should probably fall back on the long link if URL shortening fails for whatever reason.

Heufneutje reviewed (changes requested):
Just one little thing. Looks great otherwise.

HubbeKing committed:
Merge branch 'master' into generic_rss_module

HubbeKing committed:
Fix RSS parsing

lxml is the lxml HTML parser, and that's also not how beautifulsoup works anymore, like at all.

HubbeKing commented:
Should be fixed and working now, seemed to work on the testnet at least.

HubbeKing committed:
Change message-user to message-channel

It helps if you don't accidentally commit your debugging changes

HubbeKing committed:
Hey, past me, what did I JUST SAY about debugging

HubbeKing commented:
The only things remaining is that calling `.rss <feed_name>` doesn't update the feed if there's been no other messages in a channel recently, since updating is triggered by `message-channel`. This means in order to ACTUALLY get latest from the RSS feed, it'd need to be >10 minutes since the last update, and you'd need to send a non-rss command or other message first, and then do `.rss <feed_name>`

It would be handy to update feeds on something else, so there'd be updates even if channels are quiet, but then you'd need a way to specify where to send notifications of new posts.

HubbeKing committed:
Update RSS feed when added or queried, if needed

HubbeKing commented:
Okay, the module now triggers an update if needed when queried for latest item for a feed, and automatically fetches the latest item when a new feed is added.
Probably good enough? It'd be neat to update on things that aren't `message-channel`, but that's a bit more work.

HubbeKing committed:
Don't always send a response from checkFeeds

If there's no new post, there's no need.
